sockeye/knn.py

```diff
@@ -0,0 +1,175 @@
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import argparse
+from abc import abstractmethod
+from dataclasses import dataclass
+import logging
+import math
+import numpy as np
+import os
+import random
+import shutil
+from typing import Dict, Iterable, List, Optional, Tuple, Callable
+
+from . import arguments
+from sockeye import config, utils, constants as C
+from sockeye.log import setup_main_logger
+
+try:
+    import faiss
+except ImportError:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class KNNConfig(config.Config):
+    """
+    KNNConfig defines knn-specific configurations, including the information about the data store
+    as well as the index itself.
+
+    :param index_size: Size of the index and the data store.
+    :param dimension: Number of dimensions of the decoder states.
+    :param state_data_type: Data type of the decoder states (keys).
+    :param word_data_type: Data type of the stored word indexes (values).
+    :param index_type: faiss index signature, see https://github.com/facebookresearch/faiss/wiki/The-index-factory
+    :param train_data_size: Size of the training data used to train the index (if it needs to be trained).
+    """
+    index_size: int
+    dimension: int
+    state_data_type: str
+    word_data_type: str
+    index_type: str
+    train_data_size: int
+
+
+class FaissIndexBuilder:
+
+    def __init__(self, config: KNNConfig, use_gpu: bool = False, device_id: int = 0):
+        utils.init_faiss()  # faiss will definitely be used for this class, so check here
+        self.config = config
+        self.use_gpu = use_gpu
+        self.device_id = device_id
+
+    def init_faiss_index(self, train_sample: Optional[np.memmap] = None):
+        index = faiss.index_factory(self.config.dimension, self.config.index_type)
+        if self.use_gpu is True:
+            res = faiss.StandardGpuResources()
+            co = faiss.GpuClonerOptions()
+            index = faiss.index_cpu_to_gpu(res, self.device_id, index, co)
+
+        if not index.is_trained and train_sample is not None:
+            index.train(train_sample.astype(np.float32))  # unfortunately, faiss index only supports float32
+        elif not index.is_trained:
+            logger.error("Index needs training but no training sample is passed.")
+
+        return index
+
+    def add_items(self, index, keys: np.array):
+        item_count, key_dim = keys.shape
+        assert key_dim == self.config.dimension
+
+        index.add(keys.astype(np.float32))  # unfortunately, faiss index only supports float32
+
+    def block_add_items(self, index, keys: np.array, block_size: int = 1024*1024):
+        item_count, key_dim = keys.shape
+        assert key_dim == self.config.dimension
+
+        n_blocks = item_count // block_size
+        for i in range(n_blocks):
+            logger.info("adding block no.{0}".format(i))
```

Comment:
maybe we can use `logger.info("adding block no. {}", i)` or `logger.info(f'adding block no.{i}')`